### PR TITLE
Add sorting to the Queries Panel. Fix bug that truncates times over 999 ms.

### PR DESCRIPTION
**Pull request by crazyjaco**
This pull requests adds the ability to sort the queries panel by elapsed time or execution order.

It also addresses a bug from the sprintf function on the elapsed times that resulted in a time of 15,426.1 ms to be truncated to 15.0 ms. This affected any time over 999 ms as far as I could tell.

I gave @jrfnl a heads up about this pull request.

This is based on my work in this plugin:
https://github.com/crazyjaco/debug-bar-queries-plus

### CODE REVIEW

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: The `enqueue()` method is called before the panels are initialized which is why hooking in doesn't work.
https://github.com/dd32/debug-bar/blob/master/debug-bar.php#L60

I'd suggest adding the enqueue to the `Debug_Bar::enqueue()` method wrapped in the same condition as is used in the `prerender()` method: `defined( 'SAVEQUERIES' ) && SAVEQUERIES || ! empty( $GLOBALS['EZSQL_ERROR'] )`.

That conditional could possibly be abstracted to a utility method to be called both from the enqueue in the main class as well as from the prerender method here.

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: :+1:

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: Should `$counter` and `number_format...` each be wrapped in a `esc_attr()` ?

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: Not directly related to this PR, but this `translators` comment seems superfluous as there is nothing to translate on the next line.

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: I would suggest to keep the two line split as:
1. the line has now become very long, and
2. PO file generators might get confused about the two translatable strings versus the translators comment.

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: `.=` should be `=` on this first line.

This currently causes a `Notice: Undefined variable: sortby in /path/to/debug-bar/panels/class-debug-bar-queries.php on line 97` error.

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: I don't see why you'd want to use the `render_panel_info_block()` method for this. This is a different type of object, so why not just output the HTML ?

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: Personally I'm a big fan of being explicit in what you're testing for, i.e. `if ( $wpdb->num_queries > 0 )`

File: panels/class-debug-bar-panel.php
Comment (jrfnl) [inline panels/class-debug-bar-panel.php]: See my remark at the function call. I don't think this is something that belongs in this method.

File: js/debug-bar-queries.dev.js
Comment (jrfnl) [inline js/debug-bar-queries.dev.js]: Correct me if I'm wrong, but I believe line 1 and line 42 can be removed which would also allow for the nesting level to go down one for all the rest of the code.

File: panels/class-debug-bar-queries.php
Comment (jrfnl) [inline panels/class-debug-bar-queries.php]: From a usability point of view, it is counter-intuitive to have the default selection be the last in the list instead of the first.

File: js/debug-bar-queries.dev.js
Comment (jrfnl) [inline js/debug-bar-queries.dev.js]: The file needs to end with a new line.

File: js/debug-bar-queries.dev.js
Comment (jrfnl) [inline js/debug-bar-queries.dev.js]: Can the `e` variable name be made more descriptive ?

File: js/debug-bar-queries.dev.js
Comment (jrfnl) [inline js/debug-bar-queries.dev.js]: I presume you meant to compare parsedA with parsedB here and in the next condition?
If not, you can remove the parseInt statements above this line.

I'd also suggest adding a helper method for the actual compares to get rid of the duplicate code.

**Review by jrfnl — CHANGES_REQUESTED** [review]
Hi Bradley,

Looking good! I like the idea and the implementation.

I have left a number of comments inline regarding the code itself.

Please also have a look at the Travis run output - the (non-minified) JS file does not comply with the WP coding standards which is why the build is failing.

Lastly, you may also want to add some basic CSS styling for the select drop down as when the debug bar is now used on the web front-end, the styling may vary wildly depending on the theme used.

How it displays in the back-end:
screenshot

How it displays in the front-end when using twentyseventeen:
screenshot

File: debug-bar.php
Comment (jrfnl) [inline debug-bar.php]: Does this script need to be loaded in the header ? or can it be loaded after the page has rendered ?
I'd suggest setting `$in_footer` to `true`.
https://developer.wordpress.org/reference/functions/wp_enqueue_script/

File: js/debug-bar-queries.dev.js
Comment (jrfnl) [inline js/debug-bar-queries.dev.js]: Line 9-15 and 22-28 are duplicate code and could be refactored.
What about adding a helper function like:
```js
function sortQueries( $valueA, $valueB ) {
	if ( parsedA > parsedB ) {
		return 1;
	}
	if ( parsedA < parsedB ) {
		return -1;
	}
	return 0;
}
```

File: debug-bar.php
Comment (jrfnl) [inline debug-bar.php]: Is there a particular reason you choose not to implement this with a helper method ? I previously suggested this in https://github.com/dd32/debug-bar/pull/40#discussion_r132350474 ?

File: css/debug-bar.dev.css
Comment (jrfnl) [inline css/debug-bar.dev.css]: Please use a one tab indent for this line and the lines below it.

**Review by jrfnl — CHANGES_REQUESTED** [review]
Hi @crazyjaco, thanks for making these changes.
Sorry for the delay before I could get to the second review.

I'd say we're nearly there. Just a few small remarks left.
Also: could you rebase and squash the PR ? Let me know if you need help on how to do this.

I'll do another test round once you've responded to these comments and AFAICS it should be fine to merge if nothing else pops up during the test.

File: js/debug-bar-queries.dev.js
Comment (crazyjaco) [reply js/debug-bar-queries.dev.js]: Though these two pieces of code are very similar, but they are not actually the same. The comparison and return values are different.

The first instance is sorting by elapsed time so it is doing a sort of largest (slowest) down to smallest (quickest).
The second instance is an ordinal sort, sorting from smallest (first) to largest (last).

File: debug-bar.php
Comment (crazyjaco) [reply debug-bar.php]: can definitely go in footer.

File: css/debug-bar.dev.css
Comment (crazyjaco) [reply css/debug-bar.dev.css]: oops.
since this is a spacing issue, i shouldn't need to recompile the minified version, correct?

Comment (jrfnl) [reply css/debug-bar.dev.css]: Correct.

File: debug-bar.php
Comment (crazyjaco) [reply debug-bar.php]: Wasn't sure how to implement, then forgot to go back.
I have put a public helper function in the Debug_Bar class.
It seems to work well enough. Not sure if the way I called the helper in the prerender function of the queries panel is the appropriate way to do it.

File: debug-bar.php
Comment (crazyjaco) [reply debug-bar.php]: can definitely go in the footer.

File: js/debug-bar-queries.dev.js
Comment (crazyjaco) [reply js/debug-bar-queries.dev.js]: The logic in the two blocks is actually not duplicative.
The logic in the first block is for elapsed time, sorting largest (slowest) down to smallest (fastest).
The logic in the second block returns the ordinal order, smallest (first) to largest (last).

The operators and return values do not match due to this difference. That is why I did not abstract that out.

File: css/debug-bar.dev.css
Comment (crazyjaco) [reply css/debug-bar.dev.css]: oops.
since this is just a spacing issue, i shouldn't need to recompile the minified version, correct?

File: js/debug-bar-queries.dev.js
Comment (jrfnl) [reply js/debug-bar-queries.dev.js]: I completely missed that. Thanks for pointing that out.
I think it could still be refactored in that case, with one closure using return sortQueries(...); and the other return - sortQueries(...);, but I'll leave it up to you.

**Comment by crazyjaco** [comment]
Checking in on this pull request. I'm just wondering why this never got merged.

After talking with @jrfnl in September it sounded like this was good to go.
Seeing now that it has merge conflicts after sitting for so long.
Just not sure what to do to get this in.

**Comment by crazyjaco** [comment]
Resolved outstanding merge conflicts.

Looks like there was another pull request to fix a ton of codesniffer spacing issues that never got merged. I'm assuming those changes are why my build is failing in Travis.

I'm going to go ahead and try and resolve those in this request so this can be unblocked.

**Comment by crazyjaco** [comment]
Not so sure how to resolve the 3 remaining errors from the PHPCompatibility suite.

Any advice would be welcome.

**Comment by jrfnl** [comment]
@crazyjaco I'm sorry to say, but all those new changes make this PR unmergable as it is now impossible to review the actual change you want to make.

The fact that the build is not passing is outside the scope of this PR and PR #51 should have taken care of that, though that one has gone stale too by now as a new version of WPCS has been released in the mean time.

I really appreciate the amount of time & effort you have put into fixing the CS issues, but those have to be pulled in another PR. Maybe you can cherrypick them to a new branch ?
Or maybe even better, you could pull your CS updates against the `feature/update-codestyle` branch so they are added to PR #51.

**Comment by crazyjaco** [comment]
Life got in the way and I just got back to this today.
I did a git reset back to dd32's last change on my fork (after my changes, before my wpcs fixes.).
I did a git push -f to get the changes up, but it seems that somehow triggered this PR to close. I'm not really sure what just happened.

**Comment by crazyjaco** [comment]
ran
```
git reset 1deb9fc
git push -f
```

**Pull request closed without merging**
